marimo/_messaging/types.py

```diff
@@ -1,6 +1,63 @@
 # Copyright 2024 Marimo. All rights reserved.
-from typing import Any, Tuple
+import abc
+from typing import Any, Dict, Optional, Tuple
+
+from marimo._ast.cell import CellId_t
 
 # The message from the kernel is a tuple of message type
 # and a json representation of the message
 KernelMessage = Tuple[str, Any]
+
+
+class Stream(abc.ABC):
+    """
+    A stream is a class that can write messages from the kernel to
+    some output.
+    The `write` method is called by the kernel.
+    """
+
+    cell_id: Optional[CellId_t] = None
+
+    @abc.abstractmethod
+    def write(self, op: str, data: Dict[Any, Any]) -> None:
+        pass
+
+
+class Stdout:
+    @abc.abstractmethod
+    def start(self) -> None:
+        pass
+
+    @abc.abstractmethod
+    def stop(self) -> None:
+        pass
+
+    @abc.abstractmethod
+    def pause(self) -> None:
+        pass
```

Comment:
I'm not sure these (start, stop, pause) are useful methods for a generic `Stdout` type — they are sort of tied to the implementation of the pipe-based `Stdout` that we have today.

Instead, if you do want to introduce a generic abstract `Stdout` type, why not have methods that `Stdout`s must support? These would be the public methods that are on our `Stdout` implementation today.

Same for `Stderr`.

Reply:
fixed